Classes/EventListener/ProcessRequestTokenListener.php

```diff
@@ -0,0 +1,26 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Causal\Oidc\EventListener;
+
+use TYPO3\CMS\Core\Authentication\Event\BeforeRequestTokenProcessedEvent;
+use TYPO3\CMS\Core\Security\RequestToken;
+
+final class ProcessRequestTokenListener
+{
+    public function __invoke(BeforeRequestTokenProcessedEvent $event): void
+    {
+        $user = $event->getUser();
+        $requestToken = $event->getRequestToken();
+        if ($requestToken instanceof RequestToken) {
+            // fine, there is a valid request-token
+            return;
+        }
+        // @TODO how to improve security? Maybe get rid of redirects?
+        if (!isset($event->getRequest()->getQueryParams()['tx_oidc'])) {
+            return;
+        }
+        $event->setRequestToken(RequestToken::create('core/user-auth/' . strtolower($user->loginType)));
```

Comment:
Instead of creating the token here in the hook, I think it would be better to create it actually in the `AuthenticationUrlRequest` middleware. So we really leverage this security layer.

Reply:
Screw that idea, this wouldn't work of course.